qmp/libvirtgo_linux.go

```diff
@@ -0,0 +1,327 @@
+}
+
+// eventRegisterDefaultImpl registers a default event implementation.
+func eventRegisterDefaultImpl() {
+	eventRegisterID := eventRegisterDefaultImplInternal()
+	if eventRegisterID == -1 {
+		log.Printf(
+			"EventRegisterDefaultImpl returned an unexpected value %d\n",
+			eventRegisterID)
+		//TODO: panic or what???
+	}
+}
+
+var eventRegisterDefaultImplInternal = func() int {
+	return libvirt.EventRegisterDefaultImpl()
+}
+
+// domainEventRegister register with libvirt to receive events for the
+// specified domain.
+func domainEventRegister(mon *libvirtGoMonitorLinux) (int, error) {
+	callback :=
+		libvirt.DomainEventCallback(newEventCallback(mon))
+
+	//TODO: add ability to register for more event types
+	callbackID := domainEventRegisterInternal(
+		mon,
+		&callback,
+		func() {
+			// empty on purpose until it's decided
+			// what to do here.
+		},
+	)
+	if callbackID == -1 {
+		return -1, errors.New("Domain event registration failed!")
+	}
+
+	return callbackID, nil
+}
+
+var domainEventRegisterInternal = func(
+	mon *libvirtGoMonitorLinux,
+	callback *libvirt.DomainEventCallback,
+	fn func()) int {
+	return mon.virConn.DomainEventRegister(
+		*mon.domain,
+		libvirt.VIR_DOMAIN_EVENT_ID_LIFECYCLE,
+		callback,
+		fn,
+	)
+}
+
+// domainEventDeregister stops the registration with libvirt from receiving
+// events for the specified domain.
+func domainEventDeregister(
+	mon *libvirtGoMonitorLinux,
+	callbackID int) {
+	<-mon.doneChan
+	ret := domainEventDeregisterInternal(mon.virConn, callbackID)
+	if ret != 0 {
+		log.Printf("DomainEventDeregister returned an unexpected value: %d\n",
+			ret)
+	}
+}
+
+var domainEventDeregisterInternal = func(virConn *libvirt.VirConnection,
+	callbackID int) int {
+	return virConn.DomainEventDeregister(callbackID)
+}
+
+// newEventCallback convenient function to provide access
+// to the eventChan to the returned closure/callback.
+func newEventCallback(mon *libvirtGoMonitorLinux) libvirt.DomainEventCallback {
+	return func(c *libvirt.VirConnection,
+		d *libvirt.VirDomain,
+		eventDetails interface{}, f func()) int {
+
+		// if monitor is not connected,
+		// we should not continue processing messages
+		// as the mon.eventsChan will be closed.
+		if mon.virConn == nil {
+			return 0
+		}
+
+		// We only support lifecycleEvents for now
+		if lifecycleEvent, ok :=
+			eventDetails.(libvirt.DomainLifecycleEvent); ok {
+			mon.eventsChan <- constructEvent(lifecycleEvent)
+			f()
+		}
+
+		// according to the libvirt-do
+		// the return code is ignored
+		return 0
+	}
+}
+
+// pollEventsLoop keeps polling libvirt for new domain events
+func pollEventsLoop(doneChan chan bool) {
+	checkInterval := time.Tick(getPollInterval())
+	for {
+		select {
+		case <-doneChan: // stop polling for events
+			return
+		case <-checkInterval:
+			ret := eventRunDefaultImplInternal()
+			if ret == -1 {
+				doneChan <- true
+				return
+			}
+		}
+	}
+}
+
+var eventRunDefaultImplInternal = func() int {
```

Comment:
Swapping out functions during tests is messy.  Make this an interface and make implementations of that for the monitor and the tests.